spec/jobs/bulk_invoice_job_spec.rb

```diff
@@ -3,19 +3,73 @@
 require 'spec_helper'
 
 describe BulkInvoiceJob do
-  let(:order1) { build(:order, id: 1) }
-  let(:order2) { build(:order, id: 2) }
-  let(:order3) { build(:order, id: 3) }
-
-  let(:order_ids) { [3, 1, 2] }
-
   subject { BulkInvoiceJob.new(order_ids, "/tmp/file/path") }
 
   describe "#sorted_orders" do
+    let(:order1) { build(:order, id: 1) }
+    let(:order2) { build(:order, id: 2) }
+    let(:order3) { build(:order, id: 3) }
+    let(:order_ids) { [3, 1, 2] }
+
     it "returns results in their original order" do
       expect(Spree::Order).to receive(:where).and_return([order1, order2, order3])
 
       expect(subject.__send__(:sorted_orders, order_ids)).to eq [order3, order1, order2]
     end
   end
+
+  context "when invoices are enabled" do
+    before do
+      Flipper.enable(:invoices)
+    end
+
+    describe "#perform" do
+      let(:order1) { create(:order) }
+      let(:order2) { create(:order) }
+      let(:order3) { create(:order) }
+      let(:order_ids) { [order1.id, order2.id, order3.id] }
+      it "should generate invoices for invoiceable orders only" do
+        expect(subject).to receive(:sorted_orders).with(order_ids).and_return([order1, order2,
+                                                                               order3])
+        expect(order1).to receive(:invoiceable?).and_return(true)
+        expect(order2).to receive(:invoiceable?).and_return(false)
+        expect(order3).to receive(:invoiceable?).and_return(true)
+
+        [order1, order3].each do |order|
+          expect(subject).to receive(:generate_invoice).with(order)
+        end
+        expect(subject).not_to receive(:generate_invoice).with(order2)
+
+        subject.perform(order_ids, "/tmp/file/path")
+      end
+    end
+
+    describe "#generate_invoice" do
```

Comment:
Generally it's better to test the result from the public method, rather than directly testing internal methods. Perhaps it's a bit less efficient, and requires a bit more setup, but it would be less dependent on the inner workings.

Reply:
Hi @dacook 
are you suggesting removing that test?

btw the previous test `#sorted_orders` is also checking a private method.

Reply:
I suggest that everything should be tested from the public methods (`perform`) if possible.

The  "#generate_invoice" test is checking that a string from the renderer is being saved to a PDF.  I guess it's also checking that the _right_ renderer is used, but it's not using a 'real' renderer, so it doesn't seem like a thorough test.

Can we test that from `perform`? Yes I think we can check the contents of the PDF, and we could check for some orders details (like customer name and product name). To test for the _right_ renderer, is there a unique string that could be searched? If not, then yes perhaps it's worth doing the mocking here, but I'd try mocking a 'real' renderer.

Honestly I'm not sure the best way to balance between smaller/independent "unit" tests, and larger/real-life "integration" tests. But I would normally draw the line at private methods, and only use mocks sparingly.

How about we wait and see what a second reviewer says before making any change.

Reply:
I am with David here, we shouldn't be testing private method directly. As far as I understand, we are creating pdf for each invoice, so you should be able to look at the pdf to see if they have been generated with the correct data. 

We are probably in the realm of integration testing here, but I think that's fine. I am not sure if any of this would be tested somewhere else.

Reply:
Ok, so I remove the private method tests + the mocks
I added new instructions to validate the content of the generated PDF.